bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/handler/OpenThermGatewayHandler.java

```diff
@@ -108,184 +102,186 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+        scheduler.submit(() -> connectionStateChangedTask(state));
+    }
 
-                if (state != null) {
-                    logger.debug("Received update for channel '{}': {}", channelId, state);
-                    updateState(channelId, state);
-                }
+    private void connectionStateChangedTask(ConnectionState state) {
+        if (this.state != state) {
+            this.state = state;
+
+            switch (state) {
+                case INITIALIZING:
+                    updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, "Initializing");
+                    break;
+                case DISCONNECTED:
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Disconnected");
+                    autoReconnect();
+                    break;
+                case CONNECTING:
+                    updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, "Connecting");
+                    break;
+                case CONNECTED:
+                    updateStatus(ThingStatus.ONLINE);
+                    cancelAutoReconnect();
+                    break;
             }
         }
     }
 
     @Override
     public void handleRemoval() {
-        logger.debug("Removing OpenTherm Gateway handler");
+        logger.debug("Removing OpenThermGateway handler");
         disconnect();
         super.handleRemoval();
     }
 
     @Override
     public void dispose() {
+        logger.debug("Disposing OpenThermGateway handler");
         disconnect();
-
-        ScheduledFuture<?> localReconnectTask = reconnectTask;
-        if (localReconnectTask != null) {
-            localReconnectTask.cancel(true);
-            reconnectTask = null;
-        }
-
         super.dispose();
     }
 
     private void connect() {
         @Nullable
-        OpenThermGatewayConfiguration conf = config;
+        OpenThermGatewayConfiguration config = configuration;
 
-        explicitDisconnect = false;
-
-        if (connecting) {
-            logger.debug("OpenTherm Gateway connector is already connecting ...");
+        if (this.state == ConnectionState.CONNECTING) {
+            logger.debug("OpenThermGateway connector is already connecting");
             return;
         }
 
+        // Make sure everything is cleaned up before creating a new connection
         disconnect();
 
-        if (conf != null) {
-            logger.debug("Starting OpenTherm Gateway connector");
+        if (config != null) {
+            connectionStateChanged(ConnectionState.INITIALIZING);
+
+            logger.debug("Starting OpenThermGateway connector");
+
+            autoReconnect = true;
 
-            connector = new OpenThermGatewaySocketConnector(this, conf.ipaddress, conf.port);
+            connector = new OpenThermGatewaySocketConnector(this, config);
 
             Thread thread = new Thread(connector, "OpenTherm Gateway Binding - socket listener thread");
             thread.setDaemon(true);
             thread.start();
 
```

Comment:
Scheduler should only handle short time thread. If your thread is a long time thread, it must not be started by scheduler (as it will lock an entry in the thread pool.
Also note that the thread pool name does not respect our convention.

Reply:
You could do `new NamedThreadFactory("binding-xyz")` to create a custom thread. This way, the thread will be automatically named correctly.

Reply:
Done